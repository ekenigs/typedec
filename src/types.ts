export enum Types {
    string = "string",
    number = "number",
    boolean = "boolean",
    symbol = "symbol",
    enum = "enum",
    instance = "instance",
}

export type StringOptions = {};
export type BooleanOptions = {};
export type NumberOptions = {};
export type SymbolOptions = {};
export type EnumOptions = {};
export type InstanceOptions = {new(...params: any[]): {}};
export type Options = StringOptions
    | BooleanOptions
    | NumberOptions
    | SymbolOptions
    | EnumOptions
    | InstanceOptions;

export interface IConfig {
    type: Types;
    options: Options;
}

export interface IMetadata {
    [key: string]: IConfig;
}

export type TypedecDecorator = <T extends {new(...params: any[]): {}}>(constructor: T) => T;
export type TypeDecorator = (target: any, property?: string, index?: number) => void;
export type TypeDecoratorFactory<O extends Options> = (options?: O) => TypeDecorator;
