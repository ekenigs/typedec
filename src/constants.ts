export const TYPEDEC_METADATA: symbol = Symbol("TYPEDEC_METADATA");
