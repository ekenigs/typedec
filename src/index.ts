export { typedec } from "./typedec";
export { tString, tBoolean, tNumber, tSymbol, tEnum, tInstance } from "./decorators";
