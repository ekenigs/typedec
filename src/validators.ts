import { Types, IMetadata, IConfig } from "./types";

// TODO:
// improve error messages. e.g. show property name
// improve stacktrace
export function validateString(value: any): void {
    if (typeof value !== "string") {
        throw new TypeError(`must be a string`);
    }
}

export function validateBoolean(value: any): void {
    if (typeof value !== "boolean") {
        throw new TypeError(`must be an boolean`);
    }
}

export function validateNumber(value: any): void {
    if (typeof value !== "number") {
        throw new TypeError(`must be a number`);
    }
}

export function validateSymbol(value: any): void {
    if (typeof value !== "symbol") {
        throw new TypeError(`must be a symbol`);
    }
}

export function validateEnum(value: any, options: any[]): void {
    if (!options.includes(value)) {
        throw new TypeError(`must be on of ${options}`);
    }
}

export function validateInstance(value: any, options: any): void {
    if (!value || !value.constructor || !value.constructor.name || value.constructor.name !== options) {
        throw new TypeError(`must be an instance of ${options}`);
    }
}

export function validateType(config: IConfig, value: any): void {
    switch (config.type) {
        case Types.string:
            validateString(value);
            break;

        case Types.boolean:
            validateBoolean(value);
            break;

        case Types.number:
            validateNumber(value);
            break;

        case Types.symbol:
            validateSymbol(value);
            break;

        case Types.enum:
            validateEnum(value, config.options as any);
            break;

        case Types.instance:
            validateInstance(value, config.options);
            break;

        default:
            break;
    }
}

export function validateProperty(metadata: IMetadata, property: string, value: any): void {
    const config: IConfig | undefined = metadata[property];

    if (!config) {
        return;
    }

    validateType(config, value);
}

export function validateParameters(metadata: IMetadata, parameters: any[]): void {
    parameters.forEach((value: any, index: number) => {
        validateProperty(metadata, index.toString(), value);
    });
}
