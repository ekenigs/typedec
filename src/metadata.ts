import "reflect-metadata";

import { IMetadata } from "./types";
import { TYPEDEC_METADATA } from "./constants";

export function getMetadata(target: any): IMetadata {
    return Reflect.getMetadata(TYPEDEC_METADATA, target) || {};
}

export function addMetadata(target: any, key: string, value: any): void {
    const metadata: IMetadata = {
        ...getMetadata(target),
        [key]: value,
    };

    Reflect.defineMetadata(TYPEDEC_METADATA, metadata, target);
}
