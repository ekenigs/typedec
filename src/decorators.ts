import { addMetadata } from "./metadata";
import {
    Types,
    Options,
    StringOptions,
    BooleanOptions,
    NumberOptions,
    SymbolOptions,
    EnumOptions,
    InstanceOptions,
    TypeDecorator,
    TypeDecoratorFactory,
} from "./types";

function createTypeDecoratorFactory<O extends Options>(type: Types): TypeDecoratorFactory<O> {
    return <T>(options: T): TypeDecorator => {
        return (target: any, property?: string, index?: number): void => {
            const metadata: any = { type, options };

            if (type === Types.enum) {
                metadata.options = Object.values(options);
            }

            if (type === Types.instance) {
                if (!options || typeof options !== "function" || !(options as any).name) {
                    throw new TypeError("InstanceOptions must be a class");
                }

                metadata.options = (options as any).name;
            }

            if (typeof index === "number") {
                addMetadata(target, index.toString(), metadata);
            } else {
                addMetadata(target.constructor, property as string, metadata);
            }
        };
    };
}

export const tString: TypeDecoratorFactory<StringOptions> = createTypeDecoratorFactory(Types.string);
export const tBoolean: TypeDecoratorFactory<BooleanOptions> = createTypeDecoratorFactory(Types.boolean);
export const tNumber: TypeDecoratorFactory<NumberOptions> = createTypeDecoratorFactory(Types.number);
export const tSymbol: TypeDecoratorFactory<SymbolOptions> = createTypeDecoratorFactory(Types.symbol);
export const tEnum: TypeDecoratorFactory<EnumOptions> = createTypeDecoratorFactory(Types.enum);
export const tInstance: TypeDecoratorFactory<InstanceOptions> = createTypeDecoratorFactory(Types.instance);
