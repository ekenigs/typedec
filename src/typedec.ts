import { IMetadata, TypedecDecorator } from "./types";
import { getMetadata } from "./metadata";
import { validateProperty, validateParameters } from "./validators";

function createProxy(metadata: IMetadata, object: any): ProxyConstructor {
    return new Proxy(object, {
        set(target: any, property: string, value: any): boolean {
            validateProperty(metadata, property, value);

            target[property] = value;

            return true;
        },
    });
}

export function typedec(): TypedecDecorator {
    return <T extends {new(...params: any[]): {}}>(constructor: T): T => {
        return class extends constructor {
            constructor(...params: any[]) {
                const metadata: IMetadata = getMetadata(constructor);

                validateParameters(metadata, params);
                super(...params);

                return createProxy(metadata, this);
            }
        };
    };
}

export default typedec;
