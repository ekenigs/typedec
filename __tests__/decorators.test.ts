import { getMetadata } from "../src/metadata";
import { tString, tBoolean, tNumber, tSymbol, tEnum, tInstance } from "../src/decorators";

describe("decorators", () => {
    it("tString", () => {
        class TStringClass {
            @tString() public testString: string;
            constructor(@tString() testString: string) {
                this.testString = testString;
            }
        }

        expect(getMetadata(TStringClass)).toMatchSnapshot();
    });

    it("tBoolean", () => {
        class TBooleanClass {
            @tBoolean() public testBool: boolean;
            constructor(@tBoolean() testBool: boolean) {
                this.testBool = testBool;
            }
        }

        expect(getMetadata(TBooleanClass)).toMatchSnapshot();
    });

    it("tNumber", () => {
        class TNumberClass {
            @tNumber() public testNumber: number;
            constructor(@tNumber() testNumber: number) {
                this.testNumber = testNumber;
            }
        }

        expect(getMetadata(TNumberClass)).toMatchSnapshot();
    });

    it("tSymbol", () => {
        class TSymbolClass {
            @tSymbol() public testSymbol: symbol;
            constructor(@tSymbol() testSymbol: symbol) {
                this.testSymbol = testSymbol;
            }
        }

        expect(getMetadata(TSymbolClass)).toMatchSnapshot();
    });

    it("tEnum", () => {
        enum Enum {
            enum1 = "enum1",
            enum2 = "enum2",
        }
        class TEnumClass {
            @tEnum(Enum) public testEnum: any;
            constructor(@tEnum(Enum) testEnum: any) {
                this.testEnum = testEnum;
            }
        }

        expect(getMetadata(TEnumClass)).toMatchSnapshot();
    });

    describe("tInstance", () => {
        it("tInstance", () => {
            class TestClass {}
            class TInstanceClass {
                @tInstance(TestClass) public testInstance: any;
                constructor(@tInstance(TestClass) testInstance: any) {
                    this.testInstance = testInstance;
                }
            }

            expect(getMetadata(TInstanceClass)).toMatchSnapshot();
        });

        it("should throw an error because no options", () => {
            expect(() => {
                class TInstanceClass {
                    @tInstance() public testInstance: any;
                }

                return TInstanceClass;
            }).toThrowError("InstanceOptions must be a class");
        });

        it("should throw an error because options not a class", () => {
            expect(() => {
                class TInstanceClass {
                    @tInstance({} as any) public testInstance: any;
                }

                return TInstanceClass;
            }).toThrowError("InstanceOptions must be a class");
        });

        it("should throw an error because options has no a name", () => {
            expect(() => {
                class TInstanceClass {
                    @tInstance((() => true) as any) public testInstance: any;
                }

                return TInstanceClass;
            }).toThrowError("InstanceOptions must be a class");
        });
    });
});
