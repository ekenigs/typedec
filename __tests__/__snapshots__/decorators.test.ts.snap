// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators tBoolean 1`] = `
Object {
  "0": Object {
    "options": undefined,
    "type": "boolean",
  },
  "testBool": Object {
    "options": undefined,
    "type": "boolean",
  },
}
`;

exports[`decorators tEnum 1`] = `
Object {
  "0": Object {
    "options": Array [
      "enum1",
      "enum2",
    ],
    "type": "enum",
  },
  "testEnum": Object {
    "options": Array [
      "enum1",
      "enum2",
    ],
    "type": "enum",
  },
}
`;

exports[`decorators tInstance tInstance 1`] = `
Object {
  "0": Object {
    "options": "TestClass",
    "type": "instance",
  },
  "testInstance": Object {
    "options": "TestClass",
    "type": "instance",
  },
}
`;

exports[`decorators tNumber 1`] = `
Object {
  "0": Object {
    "options": undefined,
    "type": "number",
  },
  "testNumber": Object {
    "options": undefined,
    "type": "number",
  },
}
`;

exports[`decorators tString 1`] = `
Object {
  "0": Object {
    "options": undefined,
    "type": "string",
  },
  "testString": Object {
    "options": undefined,
    "type": "string",
  },
}
`;

exports[`decorators tSymbol 1`] = `
Object {
  "0": Object {
    "options": undefined,
    "type": "symbol",
  },
  "testSymbol": Object {
    "options": undefined,
    "type": "symbol",
  },
}
`;
